src/main/java/com/endava/internship/service/UserServiceImpl.java

```diff
@@ -4,66 +4,84 @@
 import com.endava.internship.domain.User;
 import com.endava.internship.service.UserService;
 
-import java.util.List;
-import java.util.Map;
-import java.util.Optional;
+import java.util.*;
 import java.util.function.Function;
 import java.util.function.Predicate;
+import java.util.stream.Collectors;
 
 public class UserServiceImpl implements UserService {
 
     @Override
     public List<String> getFirstNamesReverseSorted(List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().map(User::getFirstName).
+                sorted(Comparator.reverseOrder()).collect(Collectors.toList());
     }
 
     @Override
     public List<User> sortByAgeDescAndNameAsc(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().sorted(Comparator.comparing(User::getAge).reversed().
+                        thenComparing(User::getFirstName)).
+                collect(Collectors.toList());
     }
 
     @Override
     public List<Privilege> getAllDistinctPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().flatMap(e -> e.getPrivileges().
+                stream()).collect(Collectors.toList());
     }
 
     @Override
     public Optional<User> getUpdateUserWithAgeHigherThan(final List<User> users, final int age) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().filter(user -> user.getAge() > age || user.getPrivileges().
+                        equals(Privilege.UPDATE)).findFirst();
     }
 
     @Override
     public Map<Integer, List<User>> groupByCountOfPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().collect(Collectors.
+                groupingBy(user -> user.getPrivileges().size()));
     }
 
     @Override
     public double getAverageAgeForUsers(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().mapToDouble(User::getAge).
+                average().orElse(-1);
     }
 
     @Override
     public Optional<String> getMostFrequentLastName(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().map(User::getLastName).
+                collect(Collectors.groupingBy(Function.identity(), Collectors.counting())).
+                entrySet().stream().filter(e -> e.getValue() >= 2).
```

Comment:
why do we need `filter(e -> e.getValue() >= 2)` . If we get a list of users with only one user, we will get his LastName as most frequent one ?

Reply:
ahh ok, I saw the test